man/man1/hackdens.1
.TH HACKDENS 1NEMO "18 May 2025"

.SH "NAME"
hackdens \- local density estimator using tree algorithm

.SH "SYNOPSIS"
\fBhackdens in=\fPsnapshot \fBout=\fPimage [parameter=value]

.SH "DESCRIPTION"
\fIhackdens\fP calculates the local density in the configuration space
using the hierarchical
N-body algorithm (J. Barnes and P. Hut, \fINature\fP \fB324\fP, 446).
For small N or systems with complex geometry \fBsnapdens(1NEMO)\fP may
be preferred. For even more performance, \fBdensity(1falcON)\fP can be
used.
.PP
The density can be writtin in the slot normally used for Potentials (the default),
or if \fBwrite_at_phi=f\fP be written to a tag named \fIDensity\fP.

.SH "PARAMETERS"
.so man1/parameters
.TP 24
\fBin\fP=\fIin-file\fP
Input data will be read from \fIin-file\fP in
snapshot format.
.TP
\fBout\fP=\fIout-file\fP
Output data are written to \fIout-file\fP in snapshot format.
.TP
\fBneib\fP=\fIvalue\fP
Number of neighbors used to determine the local density. The default
of 6 is the recommended value by Casertano & Hut (1985)
[default: 6].
.TP
\fBrneib\fP=\fIvalue\fP
Initial radius to search the neighbors. \fIHackdens\fP adaptively
change this search radius during the calculation. Therefore it is
definitely required that particles in the snapshot are arranged so
that the neighboring particles have similar local density. (See NOTES)
[default: 0.1].
.TP
\fBwrite_at_phi\fP=\fIt|f\fP
Logical if the density data is written in the "Potential" slot of the
snapshot file. If false, the density will be written with tag "Density".
[default: f]
.TP
\fBrsize\fP=\fIvalue\fP
Intial size of the root box. Box will automatically expanded to keep
all particles in it [defaults:4.0]
.TP
\fBrmin\fP=\fIvalue\fP
If given, the coordinates of the lower left corner of the initial box.
If not given, the center of the root box will lie at the origin.
[defaults:none]
.TP
\fBoptions\fP=\fIoption-string\fP
Output control options, specified as a comma-separated list
of keywords.
Currently recognized keywords are:
\fBmass\fP, \fBphase\fP: output mass, phase space
data with major data outputs [defaults:mass,phase].
.TP
\fBfcells\fP=\fIvalue\fP
Ratio of cells to bodies, used  when
allocating cells [defaults:0.9].
.TP
\fBnudge\fP=\fIsize\fP
nudge overlapping particles with this positional dispersion [0]
.TP
\fBverbose=t|f\fP
Logical if print out the number of particles processed during the
calculation [defaults:f].
.TP
\fBdensity=t|f\fP
write density, or alternatively distance to Kth particle.
Default: t
.TP
\fBndim=2|3\fP
3D or 2D computation [Default: 3]
.TP
\fBdirect=t|f\fP
slower N-squared direct density computation.  Default: f
.TP
\fBnorm=0|1\fP
Mode of normalization. Earlier (2.x and before) versions would scale equal mass
systems by nbody, the current default is norm=1 which normalizes by nbody, so
densities from the slower \fPsnapdens(1NEMO)\fP agree with \fBhackdens\fP.
Default: 1


.SH "NOTES"
Since the search radius is adaptively changed during the calculation,
the local density of particles which are close in the position in the
snapshot file should be similar.
.PP
Ideally, it is best if particles are
sorted by the local density itself. Since the local density is not
available, the second choice is to sort particles with some quantity
that has strong correlation with the local density. For example, if
the potential is available, sort particles by
potential.
.PP
In many cases the system is roughly spherically symmetric.
That means the density has strong correlation with the distance from
the center of the cluster. Then sorting particles by the distance from
the center will improve the speed significantly.

.SH "EXAMPLES"
The following example takes an N-body snapshot, sort the particles
in the order of the distance from the C.M., and calculate local density.
.EX

   snapcenter nbody.dat nbody.centered
   snapsort nbody.centered nbody.sorted 
   hackdens nbody.sorted nbody.density

.EE
One can also weight the particles by its potential to some power. For this we create a Plummer sphere, without
centering, and compute the center in two ways:
.EX

   mkplummer - 100 zerocm=f | snapsort - - rank=r > p.dat
   snapcenter p.dat . report=t
   hackdens p.dat - | snapcenter  - . weight=dens report=t
   hackforce p.dat - | snapcenter - . weight="-phi*phi*phi" report=t
   hackforce p.dat - | snapcenter - . weight="phi*phi*phi*phi" report=t
   
.EE
Note that in this case the Plummer sphere should have the correct mathematical center, and not \fIre-centered\fP.
This snippet of code is expanded in a script $NEMO/scripts/csh/centering.sh, in which these different methods
can also be applied to different models, where some are expanded in the table below:
.EX

    nbody=100 models:     Plummer   King1   King3   King5  DeVauc  Dehnen
    ---------------------------------------------------------------------
    halfmass radius:        0.769   0.427   0.418   0.408   1.3
    snapcenter  on m:       0.102   0.029*  0.030*  0.033   0.299   2.75
    snapcenter  on dens:    0.060   0.039   0.036   0.030   0.030*  0.074*
    snapcenter  on phi:     0.053
    snapcenter  on phi^2:   0.050*
    snapcenter  on phi^3:   0.052   0.032   0.030   0.026*  0.035   0.079
    snapcenter  on phi^4:   0.054   0.034   0.032   0.027   0.031   0.110
    snapcenterp:            0.080
    snapcenters:            0.119:
    snapmnmx on min phi:    0.122
    snapmnmx on m 50% phi:  0.058
    snapmnmx on m 33% phi:  0.063                           0.053
    snapmnmx on m 25% phi:  0.067
    rockstar on m:          0.082:
    hop on m:               0.088:
    
A (*) denotes the best method for that model. Values scale as 1/sqrt(nbody)


.EE

.fi
Although for a Plummer sphere the best method is more than twice as good than the worst method,
this does not apply to other models. A normal mass-weighted centering for Dehnen and deVaucouleurs
models is quite bad.
.PP
\fIdensity(1falcON)\fP is considerable faster, about twice with their K=32 default. 
.EX

mkplummer - 100000 |\\
 snapscale - - mscale=10 |\\
 snapsort - - |\\
 density - - give=mxvr |\\
 snapshell - '10**[-4:4:0.1]' dens  |\\
 tabmath - - 'log(%1),log(%4)' all |\\
 tabplot - 1 2 -4 4 -14 2  line=1,1 point=2,0.1

vs.

mkplummer - 100000 |\\
 snapscale - - mscale=10 |\\
 snapsort - - |\\
 hackdens - - |\\
 snapshell - '10**[-4:4:0.1]' dens  |\\
 tabmath - - 'log(%1),log(%4)' all |\\
 tabplot - 1 2 -4 4 -14 2  line=1,1 point=2,0.1

.EE

.SH "BUGS"
The local density is calculated using (neib-1)th neighbor.
Calculation can become very slow if the input snapshot is not
well-formed. There should exist an option that forces the density will be
written in the "Aux" slot of the snapshot file. Current version does
not use standard \fIget_snap/put_snap\fP macros. KEY and AUX will be lost from
the output snapshot.
.PP
The density in \fIsnapdens(1NEMO)\fP and \fIdensity(1falcON)\fP
are in the correct units, but
in  \fIhackdens(1NEMO)\fP seem to be off by \fBnbody\fP for equal mass
particles. For example, a Plummer sphere in virial (nbody) units should have
a central density
.nf
    (32/(3.pi^2))^2 ~ 1.168
.fi

.SH "SEE ALSO"
hackforce(1NEMO), snapdens(1NEMO), snapcenterp(1NEMO), snapcenters(1NEMO), density(1falcON), snapshell(1NEMO)
.PP
https://ui.adsabs.harvard.edu/abs/1985ApJ...298...80C/ - Casertano & Hut (1985)

.SH "ADS"
@ads 1985ApJ...298...80C


.SH "TIMING"
N=4096, merger remnant
.PP
Presorted snapshot: 5 minutes on Sun-3/60, particles are presorted by the distance from
the center of mass of the system.
.PP
Random snapshot: 60 minutes Sun-3/60

.SH "AUTHOR"
Jun Makino

.SH "HISTORY"
.nf
.ta +1.5i +4.5i
2-Dec-88  	V1.0:Created  	Jun
6-jul-89	doc updated?	Jun
23-oct-90	doc updated	Peter
18-jul-92	printf -> dprintf to make it pipable	Peter
24-may-02	fixed running out of bits for large-N 	PJT
11-aug-2022	added example hackforce vs. hackdens	PJT
21-sep-2023	added direct=	PJT
11-oct-2023	V3.0 added norm=1 as a new default	PJT
12-oct-2023	V3.0 proper mass/nbody scaling 	PJT
.fi


src/nbody/trans/hackdens/hackdens.c
/*
 * HACKDENS.C: hackcode local density calculation tool.
 *
 *          1988  V1.0   originally written        Jun Makino
 *	7-jul-89  V1.1   updated with get_history()	PJT
 *     23-oct-90  V1.2   helpvec			PJT
 *     18-jul-92  V1.3   replaced many if(debug)printf(...) by dprintf(1,...)
 *			 Also nemo_main() and usage	PJT
 *	1-apr-01      a  compiler warning
 *     24-may-02  V2.0   fixed for high-N system by using int_hack in load.c
 *     29-may-02  V2.1   add nudge= keyword to nudge overlapping particles
 *     25-apr-06  V2.2b  use global to isolate extern's (for Mac linking)
 *     28-jul-06  V2.2c  default for tag is now Density
 *                V2.2d  clarify D vs. P, working with std snapshot, not archaic
 *     21-dep-23  V2.3   add a slow direct= for benchmark/comparison
 *     12-oct-23  V2.4   scale by mass
 *     11-OCT-23  v3.0   add norm=1 and made it the default
 *
 * NOTE:   for snapshots with unequal masses this program doesn't work
 *
 */

#define global

#include "defs.h"
#include <getparam.h>
#include <history.h>
#include <filestruct.h>
#include <snapshot/snapshot.h>

string defv[] = {	
    "in=???\n			  input snapshot",
    "out=\n			  optional output file with density results ",
    "neib=6\n			  number of neighbours to define local density ",
    "rneib=0.1\n		  initial guess for neighbour sphere radius ",
    "write_at_phi=f\n		  flag to write density with Potential instead of Density tag",
    "rsize=4.0\n		  side-length of initial box",
    "rmin=\n			  lower left corner of initial box",
    "options=phase,mass\n	  misc. control options {phase, mass}",
    "fcells=0.9\n		  cell/body allocation ratio ",
    "nudge=0\n                    nudge overlapping particles with this dispersion",
    "verbose=f\n		  flag to print # of particles finished ",
    "density=t\n                  write density, or distance to Kth particle",
    "ndim=3\n                     3D or 2D computation",
    "direct=f\n                   slower direct density computation",
    "norm=1\n                     normalization mode (0=nothing   1=1/N)",
    "VERSION=3.0\n		  12-oct-2023 PJT",
    NULL,
};

string usage = "hackcode local density calculation tool";


void inputdata(void);
void readsnapshot(bodyptr *btab_ptr, int *nobj_ptr, stream instr);
void dencalc(void);
void outresult(void);
void writesnapshot(void);

// util.c
extern void pickvec(vector x, bool cf);

// load.c
extern void maketree(bodyptr btab, int nbody, double nudge);


void nemo_main()
{
    inputdata();			/* input mass and test data */
    dencalc();				/* find force at test pos */
    outresult();			/* write snap with results */
}

bodyptr massdata;		/* array of mass points */
int nmass;			/* number of mass points */
real totalmass;

bodyptr testdata;		/* array of test points */
int ntest;			/* number of test points */

bool Qdirect;
bool norm;

void inputdata()
{
    string input;
    stream instr;

    input = getparam("in");
    instr = stropen(input, "r");
    Qdensity = getbparam("density");
    Qdirect = getbparam("direct");
    readsnapshot(&massdata, &nmass, instr);	/* read mass coord data */
    strclose(instr);
    testdata = massdata;
    ntest = nmass;
    norm = getiparam("norm");
}

real tsnap;

void readsnapshot(btab_ptr, nobj_ptr, instr)
bodyptr *btab_ptr;				/* gets particle array */
int *nobj_ptr;					/* gets number of bodies */
stream instr;					/* stream to read from */
{
    int nobj, cs, i;
    real *mbuf, *mp, *pbuf, *pp;
    bodyptr bp;

    get_history(instr);

    get_set(instr, SnapShotTag);
    get_set(instr, ParametersTag);
    if (get_tag_ok(instr, TimeTag))
	get_data_coerced(instr, TimeTag, RealType, &tsnap, 0);
    else
	tsnap=0.0;
    get_data(instr, NobjTag, IntType, &nobj, 0);
    if (nobj < 1)
	error("readsnapshot: %s = %d  is absurd\n", NobjTag, nobj);
    get_tes(instr, ParametersTag);
    get_set(instr, ParticlesTag);
    get_data(instr, CoordSystemTag, IntType, &cs, 0);
    if (cs != CSCode(Cartesian, NDIM, 2))
	error("readsnapshot: cannot handle %s = %d\n", CoordSystemTag, cs);
    mbuf = mp = (real *) malloc(nobj * sizeof(real));
    pbuf = pp = (real *) malloc(nobj * 2 * NDIM * sizeof(real));
    if (mp == NULL || pp == NULL)
	error("readsnapshot: not enuf memory for buffers\n");
    get_data_coerced(instr, MassTag, RealType, mbuf, nobj, 0);
    get_data_coerced(instr, PhaseSpaceTag, RealType, pbuf, nobj, 2, NDIM, 0);
    get_tes(instr, ParticlesTag);
    get_tes(instr, SnapShotTag);
    *btab_ptr = bp = (bodyptr) malloc(nobj * sizeof(body));
    if (bp == NULL)
	error("readsnapshot: not enuf memory for bodies");
    totalmass = 0.0;
    for (i = 0; i < nobj; i++) {
	Type(bp) = BODY;
	Mass(bp) = *mp++;
	totalmass += Mass(bp);
	SETV(Pos(bp), pp);
	pp += NDIM;
	SETV(Vel(bp), pp);
	pp += NDIM;
	bp++;
    }
    dprintf(1,"Total mass = %g, nbody=%d\n", totalmass,nobj);
    free(mbuf);
    free(pbuf);
    *nobj_ptr = nobj;
}

real *dendata;			/* local density */

int n2btot, nbctot;		/* body-body, body-cell interactions */

real cputree, cpufcal;		/* CPU time to build tree, compute forces */

void dencalc()
{
    real hackden(), directden();
    real *pp, *work, rneib, newrneib, nudge;
    int neibnum;
    //double cputime(), cpubase;
    double cpubase;
      // , atof();
    // string *burststring(), *rminxstr;
    string *rminxstr;
    int xstrlen(), i, ibody;
    bodyptr bp;
    bool verbose;

    verbose=getbparam("verbose");
    rneib=getdparam("rneib");
    neibnum=getiparam("neib")+1;
    nudge = getdparam("nudge");
    if (nudge > 0) {
      set_xrandom(0);   /* should use seed= */
#if 0
      warning("Nudging all particles by +/-%s",nudge);
      for (bp=massdata; bp<massdata+nmass; bp++) {
	for (i=0; i<NDIM; i++)
	  Pos(bp)[i] += xrandom(-nudge,nudge);
      }
#endif
    }
#ifdef DEBUG
    dprintf(0,"neib=%d rneib=%f\n", neibnum, rneib);
#endif    
    rsize = getdparam("rsize");
    rminxstr = burststring(getparam("rmin"), ", ");
    if (xstrlen(rminxstr, sizeof(string)) < NDIM) {
	SETVS(rmin, - rsize / 2.0);
    } else
	for (i = 0; i < NDIM; i++)
	    rmin[i] = atof(rminxstr[i]);
    dprintf(0,"initial rsize: %8f    rmin: %8f  %8f  %8f\n",
	   rsize, rmin[0], rmin[1], rmin[2]);
    fcells = getdparam("fcells");
    dendata = pp = (real *) malloc(ntest * sizeof(real));
    work = (real *) malloc(ntest * sizeof(real));
    if (pp == NULL || work==NULL)
	error("forcecalc: not enuf memory for results");
    cpubase = cputime();
    maketree(massdata, nmass,nudge);
    cputree = cputime() - cpubase;
    dprintf(0,"  final rsize: %8f    rmin: %8f  %8f  %8f\n",
	   rsize, rmin[0], rmin[1], rmin[2]);
    cpubase = cputime();
    n2btot = nbctot = 0;
    ibody=0;
    for (bp = testdata; bp < testdata+ntest; bp++) {
	if (Qdirect) 
	  *pp = directden(bp, neibnum, rneib, work, testdata, ntest);
	else
	  *pp = hackden(bp, neibnum, rneib, &newrneib, work);
	rneib=0.95*rneib+0.05*newrneib;
	pp++;
	ibody++;
	if(verbose && ibody%100==0)dprintf(0," %d rn=%f\n", ibody,rneib);
    }
    cpufcal = cputime() - cpubase;
    if (norm==1) {
      real factor = 1.0/ntest;
      if (Qdensity) factor *= totalmass;
      dprintf(0,"Renormalizing %d densities by %g\n",ntest,factor);
      for (pp=dendata; pp < dendata+ntest; pp++) 
	*pp *= factor;
    }
}

stream outstr;

void outresult()
{
    string out;

    out = getparam("out");
    if (*out != 0) {
	outstr = stropen(out, "w");
	put_history(outstr);
	writesnapshot();			/* output testdata results */
	strclose(outstr);
    }
}

void writesnapshot()
{
    real *mbuf, *mp, *pspbuf, *pspp;
    bodyptr bp;
    int cs = CSCode(Cartesian, NDIM, 2);
    string options = getparam("options");
    bool scanopt();

    mbuf = mp = (real *) malloc(ntest * sizeof(real));
    pspbuf = pspp = (real *) malloc(ntest * 2 * NDIM * sizeof(real));
    if (mbuf == NULL || pspbuf == NULL)
	error("writesnapshot: not enuf memory for particle buffers");
    for (bp = testdata; bp < testdata+ntest; bp++) {
	*mp++ = Mass(bp);
	SETV(pspp, Pos(bp));
	pspp += NDIM;
	SETV(pspp, Vel(bp));
	pspp += NDIM;
    }
    put_set(outstr, SnapShotTag);
    put_set(outstr, ParametersTag);
    put_data(outstr, NobjTag, IntType, &ntest, 0);
    put_data(outstr, TimeTag, RealType, &tsnap, 0);
    put_tes(outstr, ParametersTag);
    put_set(outstr, ParticlesTag);
    put_data(outstr, CoordSystemTag, IntType, &cs, 0);
    if (scanopt(options, "mass"))
	put_data(outstr, MassTag, RealType, mbuf, ntest, 0);
    if (scanopt(options, "phase"))
	put_data(outstr, PhaseSpaceTag, RealType, pspbuf, ntest, 2, NDIM, 0);
    put_data(outstr, 
	     getbparam("write_at_phi") ? PotentialTag : DensityTag,
	     RealType, dendata, ntest, 0);
    put_tes(outstr, ParticlesTag);
    put_set(outstr, DiagnosticsTag);
    put_data(outstr, "cputree", RealType, &cputree, 0);
    put_data(outstr, "cpucal", RealType, &cpufcal, 0);
    put_tes(outstr, DiagnosticsTag);
    put_tes(outstr, SnapShotTag);
    free(mbuf);
    free(pspbuf);
}
